File: Scala/test-3/PalindromeTree.scala
object PalindromeTree {
  val EertreeAlpha = 26

  case class Node(
    len: Int,
    suffixLink: Int,
    children: Vector[Int],
    cnt: Long,
    endPos: Int
  )

  case class Eertree(
    nodes: Vector[Node],
    size: Int,
    last: Int,
    str: String,
    strLen: Int
  )

  def eertreeBuild(s: Option[String], len: Int): Option[Eertree] =
    sys.error("PalindromeTree.eertreeBuild not implemented")

  def eertreeCountDistinct(t: Option[Eertree]): Int =
    sys.error("PalindromeTree.eertreeCountDistinct not implemented")

  def eertreeCountOccurrences(t: Option[Eertree], pattern: Option[String], plen: Int): Long =
    sys.error("PalindromeTree.eertreeCountOccurrences not implemented")

  def eertreeLongestLength(t: Option[Eertree]): Int =
    sys.error("PalindromeTree.eertreeLongestLength not implemented")
}

File: Java/test-3/PalindromeTree.java
public abstract class PalindromeTree {
    protected PalindromeTree() {
    }

    public static final int EERTREE_ALPHA = 26;

    public static final class Node {
        public int len;
        public int suffixLink;
        public int[] children;
        public long cnt;
        public int endPos;

        public Node() {
            this.children = new int[EERTREE_ALPHA];
        }
    }

    public static final class Eertree {
        public Node[] nodes;
        public int size;
        public int capacity;
        public int last;
        public String str;
        public int strLen;
    }

    public abstract Eertree eertreeBuild(String s, int len);

    public abstract int eertreeCountDistinct(Eertree t);

    public abstract long eertreeCountOccurrences(Eertree t, String pattern, int plen);

    public abstract int eertreeLongestLength(Eertree t);
}
